### PR TITLE
DBZ-2543 Initial OpenLogReplicator adapter

### PR DESCRIPTION
**Pull request by Naros**
https://issues.redhat.com/browse/DBZ-2543

## Tasks for Debezium for Oracle:
- [x] Improve SCN restart/recovery (may require OpenLogReplicator change)
- [x] ~~Generate OpenLogReplicator configuration?~~ Document OpenLogReplicator
- [x] ~~Integrate make-shift value converter code with `OracleValueConveters`~~
- [x] Connector adapter metrics
- [x] Connect-based temporal precision not supported
- [x] Make source, host, and port configurable.

## Tasks for OpenLogReplicator:
- [x] Payload schema does not include catalog name
- [x] Create table schema changes have no schema details
- [x] Timestamps with timezone details not supported 
- [x] Timestamps with local time zone not supported
- [x] ~~ROWID data type not supported~~ (disabled tests)
- [x] Interval data types not supported 
- [x] ~~Expected sequence `\u0003` and `\u0001` in certain schema change events~~

## Follow-up Jiras Required
- [x] XML data type not supported (OLR) (https://issues.redhat.com/browse/DBZ-6896)
- [x] Capture all schema changes, not just table-specific schema changes? (https://issues.redhat.com/browse/DBZ-6897)
- [x] OpenLogReplicator baked into Oracle images (https://issues.redhat.com/browse/DBZ-6898)
- [x] Refactor Oracle Streaming MBean implementations (https://issues.redhat.com/browse/DBZ-6899)
- [x] Support mid-transaction restart recovery (https://issues.redhat.com/browse/DBZ-6895)
- [x] Confirmation problems (https://issues.redhat.com/browse/DBZ-6895)

### CODE REVIEW

**Comment by Naros** [comment]
Excluding XML -- tests look great!
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Debezium Build Aggregator 2.4.0-SNAPSHOT:
[INFO] 
[INFO] Debezium Build Aggregator .......................... SUCCESS [  0.409 s]
[INFO] Debezium IDE Formatting Rules ...................... SUCCESS [  0.437 s]
[INFO] Debezium Checkstyle Rules .......................... SUCCESS [  0.731 s]
[INFO] Debezium Revapi Rules .............................. SUCCESS [  0.036 s]
[INFO] Debezium Parent POM ................................ SUCCESS [  0.617 s]
[INFO] Debezium API ....................................... SUCCESS [  4.274 s]
[INFO] Debezium Core ...................................... SUCCESS [01:03 min]
[INFO] Debezium ANTLR DDL parsers ......................... SUCCESS [ 35.635 s]
[INFO] Debezium Assembly Descriptors ...................... SUCCESS [  0.074 s]
[INFO] Debezium Embedded .................................. SUCCESS [05:12 min]
[INFO] Debezium Storage Module ............................ SUCCESS [  0.017 s]
[INFO] Debezium Storage Kafka Module ...................... SUCCESS [  0.349 s]
[INFO] Debezium Storage File Module ....................... SUCCESS [  0.181 s]
[INFO] Debezium Schema Generator .......................... SUCCESS [  1.184 s]
[INFO] Debezium Connector for Oracle ...................... SUCCESS [25:06 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  32:06 min
[INFO] Finished at: 2023-08-10T16:41:16-04:00
[INFO] ------------------------------------------------------------------------
```

File: debezium-connector-oracle/pom.xml [hunk 1]
```diff
@@ -70,6 +70,22 @@
             <artifactId>xmlparserv2</artifactId>
             <scope>provided</scope>
         </dependency>
+        <!-- Dependencies for OpenLogReplicator -->
+        <dependency>
+            <groupId>io.vertx</groupId>
+            <artifactId>vertx-core</artifactId>
+            <version>4.2.4</version>
```
Comment (jpechane) [inline debezium-connector-oracle/pom.xml:77]: Please eternalize the version into BOM or if not possible at least use system property.

File: debezium-connector-oracle/pom.xml [hunk 2]
```diff
@@ -270,6 +286,39 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>kr.motd.maven</groupId>
+                <artifactId>os-maven-plugin</artifactId>
```
Comment (jpechane) [inline debezium-connector-oracle/pom.xml:291]: Use BOM/system property.

File: debezium-connector-oracle/pom.xml [hunk 3]
```diff
@@ -270,6 +286,39 @@
...
+                        <goals>
+                            <goal>detect</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.xolstice.maven.plugins</groupId>
+                <artifactId>protobuf-maven-plugin</artifactId>
+                <version>0.6.1</version>
```
Comment (jpechane) [inline debezium-connector-oracle/pom.xml:305]: See above

File: debezium-connector-oracle/pom.xml [hunk 4]
```diff
@@ -270,6 +286,39 @@
...
+                            <goal>compile</goal>
+                            <goal>test-compile</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <additionalProtoPathElements>
+                        <additionalProtoPathElement>${project.basedir}/src/main/proto</additionalProtoPathElement>
+                    </additionalProtoPathElements>
+                    <protocArtifact>com.google.protobuf:protoc:3.22.3:exe:${os.detected.classifier}</protocArtifact>
```
Comment (jpechane) [inline debezium-connector-oracle/pom.xml:319]: Can't we use the same plugin we do for PostgreSQL?

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OpenLogReplicatorIT.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    @After
+    public void after() throws SQLException {
+        stopConnector();
+        if (connection != null && connection.isConnected()) {
+            connection.close();
+        }
+    }
+
+    @Test
+    public void shouldStreamChanges() throws Exception {
```
Comment (jpechane) [inline debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OpenLogReplicatorIT.java:51]: Is this running for all adapters?

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleBlobDataTypesIT.java [hunk 6]
```diff
@@ -950,6 +952,66 @@ public void shouldNotStreamAnyChangesWhenLobEraseIsDetected() throws Exception {
         assertNoRecordsToConsume();
     }
 
+    @Test
+    @FixFor({ "DBZ-2948", "DBZ-5773" })
+    @SkipWhenAdapterNameIsNot(value = SkipWhenAdapterNameIsNot.AdapterName.OLR, reason = "OpenLogReplicator does not differentiate between LOB operations")
+    public void shouldStreamChangesWhenLobEraseIsDetected() throws Exception {
+        String ddl = "CREATE TABLE BLOB_TEST ("
```
Comment (jpechane) [inline debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleBlobDataTypesIT.java:959]: Is this suppurted for XStream?

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleClobDataTypeIT.java [hunk 7]
```diff
@@ -1114,14 +1116,73 @@ public void shouldNotStreamAnyChangesWhenLobEraseIsDetected() throws Exception {
...
         Awaitility.await().atMost(Duration.ofMinutes(1))
                 .until(() -> logminerLogInterceptor.containsWarnMessage("LOB_ERASE for table")
                         || xstreamLogInterceptor.containsWarnMessage("LOB_ERASE for table"));
         assertNoRecordsToConsume();
     }
 
+    @Test
+    @FixFor({ "DBZ-2948", "DBZ-5773" })
+    @SkipWhenAdapterNameIsNot(value = SkipWhenAdapterNameIsNot.AdapterName.OLR, reason = "OpenLogReplicator does not differentiate between LOB operations")
+    public void shouldStreamChangesWhenLobEraseIsDetected() throws Exception {
```
Comment (jpechane) [inline debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleClobDataTypeIT.java:1129]: Is this supported for XStream?

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/olr/OpenLogReplicatorStreamingChangeEventSource.java [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+                        offsetContext,
+                        Operation.TRUNCATE,
+                        new Object[table.columns().size()],
+                        new Object[table.columns().size()],
+                        table,
+                        schema,
+                        clock));
+    }
+
+    private Object resolveColumnValue(TableId tableId, Column column, Values values) {
```
Comment (jpechane) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/olr/OpenLogReplicatorStreamingChangeEventSource.java:476]: Should not this be handled via `OracleValueConverter` or a subcalss of it?

File: debezium-connector-oracle/src/main/proto/OraProtoBuf.proto [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+OpenLogReplicator is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with OpenLogReplicator; see the file LICENSE;  If not see
+<http://www.gnu.org/licenses/>.  */
+
+syntax = "proto3";
```
Comment (jpechane) [inline debezium-connector-oracle/src/main/proto/OraProtoBuf.proto:19]: Is protobuf even used?

**Review by jpechane — COMMENTED** [review]
@Naros First round of review done. This is very nice work! I left copuple of comments but generally the code LGTM.

File: debezium-connector-oracle/pom.xml (hunk 1 above)
Comment (Naros) [reply debezium-connector-oracle/pom.xml]: Thanks for pointing this out.  I ended up not using vert.x and there was only one use of it with `Buffer` while sending data to OpenLogReplicator.  I've replaced that code and removed vert.x as a dependency.

File: debezium-connector-oracle/pom.xml (hunk 4 above)
Comment (Naros) [reply debezium-connector-oracle/pom.xml:314]: Great idea, just tested and it works.  I've adjusted the BOM to match the same as from PostgreSQL.

File: debezium-connector-oracle/pom.xml (hunk 2 above)
Comment (Naros) [reply debezium-connector-oracle/pom.xml:286]: No longer applicable, used PG plugin instead as suggested.

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OpenLogReplicatorIT.java (hunk 5 above)
Comment (Naros) [reply debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OpenLogReplicatorIT.java]: Ah, this should not have been added, as it was an early basic test for OLR.  Removed.

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleBlobDataTypesIT.java (hunk 6 above)
Comment (Naros) [reply debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleBlobDataTypesIT.java:959]: There is now two methods:
* shouldNotStreamAnyChangesWhenLobEraseIsDetected
* shouldStreamChangesWhenLobEraseIsDetected

The former works for LogMiner/XStream and the later only works for OpenLogReplicator.  The new test added was to show that when using `DBMS_LOB.ERASE` that we cannot differentiate between the LOB operations.

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleClobDataTypeIT.java (hunk 7 above)
Comment (Naros) [reply debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleClobDataTypeIT.java:1129]: This is the same behavior as from the Blob class, there are two tests with similar names, and the new test is to show that DBMS_LOB.ERASE cannot be differentiated with OpenLogReplicator and that we get the data in its _final_ form.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/olr/OpenLogReplicatorStreamingChangeEventSource.java (hunk 8 above)
Comment (Naros) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/olr/OpenLogReplicatorStreamingChangeEventSource.java:476]: I'll see what can be done on this.  We'll definitely need to keep the `UNAVAILABLE_VALUE` loop in the event source handler as the value converters expect to look for this marker value and do the substitution; so I'll focus primarily on the _switch_ statement bit being wrapped into the value converter.

**Comment by Naros** [comment]
@jpechane I took the liberty of testing this PR merged against `main` locally for LogMiner and XStream; in both cases, the tests are all green.  The only test failures are when using the `oracle-olr` profile, but that's expected due to the current known issues with OLR.

I've merged the code and docs PRs related to DBZ-2543.  There is also a website PR (https://github.com/debezium/debezium.github.io/pull/947) adding an `OpenLogReplicator` attribute to the test matrices that should be applied just before the next release.

**Pull request merged**